src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+            inCPUImage->setBufferAndRegions(std::move(inCPUBuffer), inCpuImageRegionsDynArray);
+            assert(inCPUImage->getBuffer());
+            assert(inCPUImage->getRegions().size() > 0u);
+        }
+
+        // outCPUImage is an image matching the params of dstImage but with the extents and layer count of the current region being copied and mipLevel 1u
+        // the buffer of this image is set to (stagingBufferPointer + stagingBufferOffset) and the related region is set to cover the whole copy region (offset from 0)
+        {
+            auto outCpuImageRegionsDynArray = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1);
+            auto& outCpuImageRegion = outCpuImageRegionsDynArray->front();
+            outCpuImageRegion = {};
+            outCpuImageRegion.bufferOffset = 0u;
+            outCpuImageRegion.bufferRowLength = regionToCopyNext.bufferRowLength;
+            outCpuImageRegion.bufferImageHeight = regionToCopyNext.bufferImageHeight;
+            outCpuImageRegion.imageSubresource.aspectMask = mainRegion.imageSubresource.aspectMask;
+            outCpuImageRegion.imageSubresource.mipLevel = 0u;
+            outCpuImageRegion.imageSubresource.baseArrayLayer = 0u;
+            outCpuImageRegion.imageOffset.x = 0u;
+            outCpuImageRegion.imageOffset.y = 0u;
+            outCpuImageRegion.imageOffset.z = 0u;
+            outCpuImageRegion.imageExtent.width    = regionToCopyNext.imageExtent.width;
+            outCpuImageRegion.imageExtent.height   = regionToCopyNext.imageExtent.height;
+            outCpuImageRegion.imageExtent.depth    = regionToCopyNext.imageExtent.depth;
+            outCpuImageRegion.imageSubresource.layerCount = core::max(regionToCopyNext.imageSubresource.layerCount, 1u);
+
+            asset::ICPUImage::SCreationParams outCPUImageParams = dstImageParams;
+            uint8_t* outCpuBufferPointer = reinterpret_cast<uint8_t*>(stagingBufferPointer) + stagingBufferOffset;
+            outCPUImageParams.flags = asset::IImage::ECF_NONE; // Because we may want to write to first few layers of CUBEMAP (<6) but it's not valid to create an Cube ICPUImage with less that 6 layers.
+            outCPUImageParams.extent = regionToCopyNext.imageExtent;
+            outCPUImageParams.arrayLayers = regionToCopyNext.imageSubresource.layerCount;
+            outCPUImageParams.mipLevels = 1u;
+            outCPUImage = asset::ICPUImage::create(std::move(outCPUImageParams));
+            assert(outCPUImage);
+            core::smart_refctd_ptr<asset::ICPUBuffer> outCPUBuffer = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(outCPUBufferSize, outCpuBufferPointer, core::adopt_memory);
+            outCPUImage->setBufferAndRegions(std::move(outCPUBuffer), outCpuImageRegionsDynArray);
+            assert(outCPUImage->getBuffer());
+            assert(outCPUImage->getRegions().size() > 0u);
+        }
+    };
+
```

Comment:
below this line is a lot of "duplicate code" could you refactor into a method (or several) ?

(maybe with a dispact switch + template on the swizzle)

Reply:
If you mean the the code inside each of the 4 if statements when using the filters (starting from `if (srcImageFormat == dstImageFormat)`, I guess I can refactor them into a method

Reply:
yeah the region and filter set-up